### PR TITLE
Add Symbols end to end tests

### PR DESCRIPTION
**Pull request by shishirx34**
This PR adds end to end test for symbols. Addresses https://github.com/NuGet/NuGetGallery/issues/5520

The test basically, builds sample template project and creates nupkg and snupkg with dlls and pdbs respectively. Pushes this to gallery and then listens to the symbol server endpoint for the availability of the PDB.

Had to update the target framework to net46 due to updated client dependencies which were required to allow symbols validations using client code.

Edit: I will add more tests to this as appropriate.

### CODE REVIEW

File: src/NuGet.Services.EndToEnd/Support/Package.cs [hunk 1]
```diff
@@ -100,5 +135,19 @@ public static Package SignedPackage()
...
             return new Package(id, normalizedVersion, fullVersion, nupkgBytes);
         }
+
+        private static MemoryStream GetMemoryStreamForFile(string filename)
+        {
+            var memoryStream = new MemoryStream();
+            using (FileStream fileStream = File.OpenRead(filename))
+            {
+                memoryStream.SetLength(fileStream.Length);
+                fileStream.Read(memoryStream.GetBuffer(), 0, (int)fileStream.Length);
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Package.cs]: Can you use [`Stream.CopyTo`]([https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copyto?view=netframework-4.7.2#System_IO_Stream_CopyTo_System_IO_Stream](https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.copyto?view=netframework-4.7.2#System_IO_Stream_CopyTo_System_IO_Stream)_) instead?

File: src/NuGet.Services.EndToEnd/Support/TestData.cs [hunk 2]
```diff
@@ -16,32 +17,66 @@ public static class TestData
         public static readonly TimeSpan RegistrationWaitDuration = TimeSpan.FromMinutes(20);
         public static readonly TimeSpan SearchWaitDuration = TimeSpan.FromMinutes(35);
         public static readonly TimeSpan V3SleepDuration = TimeSpan.FromSeconds(5);
+        public static readonly TimeSpan SymbolsWaitDuration = TimeSpan.FromMinutes(20);
+        public static readonly TimeSpan SymbolsSleepDuration = TimeSpan.FromSeconds(5);
 
         public static Stream BuildPackageStream(PackageCreationContext context)
+        {
+            var files = new List<PhysicalPackageFile>() {
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/TestData.cs]: > { [](start = 55, length = 2)

New line for `{`. We should add `.editorconfig` files to our projects...

File: src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj [hunk 3]
```diff
@@ -124,6 +141,7 @@
     <ExternalConfig Include="ExternalConfig/*.*">
     </ExternalConfig>
   </ItemGroup>
+  <ItemGroup />
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj]: remove

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 4]
```diff
@@ -108,10 +108,13 @@ public async Task<IList<string>> AutocompletePackageVersionsAsync(string id, boo
             return await _httpClient.GetJsonAsync<List<string>>(uri, logger);
         }
 
-        public async Task PushAsync(Stream nupkgStream, ITestOutputHelper logger)
+        public async Task PushAsync(Stream nupkgStream, ITestOutputHelper logger, bool isSymbolsPackage = false)
         {
             var galleryEndpoint = await GetGalleryUrlAsync(logger);
-            var url = $"{galleryEndpoint}/api/v2/package";
+            var url = isSymbolsPackage
+                ? $"{galleryEndpoint}/api/v2/symbolpackage"
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Should we look up the push URLs from the service index? This is what the client does, it'd be a "stronger" test.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs [hunk 5]
```diff
@@ -108,10 +108,13 @@ public async Task<IList<string>> AutocompletePackageVersionsAsync(string id, boo
             return await _httpClient.GetJsonAsync<List<string>>(uri, logger);
         }
 
-        public async Task PushAsync(Stream nupkgStream, ITestOutputHelper logger)
+        public async Task PushAsync(Stream nupkgStream, ITestOutputHelper logger, bool isSymbolsPackage = false)
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: Interface should match implementation

Comment (joelverhagen) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: (there is no default on the interface)

File: src/NuGet.Services.EndToEnd/Support/Clients/SymbolServerClient.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+            {
+                using (var httpClient = new HttpClient())
+                {
+                    // Add Checksum headers for fetching the pdbs from symbol server.
+                    httpClient.DefaultRequestHeaders.Add(SymbolServerHeaderKey, SymbolServerHeaderValue);
+                    using (var response = await httpClient.GetAsync(symbolFileEndpoint))
+                    {
+                        complete = response.StatusCode == HttpStatusCode.OK;
+
+                        if (!complete && duration.Elapsed + TestData.SymbolsSleepDuration < TestData.SymbolsWaitDuration)
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Clients/SymbolServerClient.cs]: this delay should be outside of the usings

File: src/NuGet.Services.EndToEnd/Support/Clients/SymbolServerClient.cs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+            var complete = false;
+            var symbolFileEndpoint = new Uri(string.Format(_testSettings.SymbolServerUrlTemplate, indexedFileName));
+
+            var duration = Stopwatch.StartNew();
+            while (!complete && duration.Elapsed < TestData.SymbolsWaitDuration)
+            {
+                using (var httpClient = new HttpClient())
+                {
+                    // Add Checksum headers for fetching the pdbs from symbol server.
+                    httpClient.DefaultRequestHeaders.Add(SymbolServerHeaderKey, SymbolServerHeaderValue);
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Clients/SymbolServerClient.cs]: >SymbolServerHeaderValue [](start = 80, length = 23)

I'm not familiar with this header. Does this mean the symbol server should do checksum verification? Could you expand the comment and maybe link some documentation?

File: src/NuGet.Services.EndToEnd/Support/PackageProperties.cs [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+namespace NuGet.Services.EndToEnd.Support
+{
+    public class PackageProperties
+    {
+        public bool IsSymbolsPackage { get; }
+
+        public HashSet<string> IndexedFiles { get; }
+
+        public PackageProperties() { }
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PackageProperties.cs]: Shouldn't this default `IndexedFiles` to an empty set?

File: src/NuGet.Services.EndToEnd/Support/PackageType.cs [hunk 9]
```diff
@@ -19,5 +19,6 @@ public enum PackageType
         SemVer2StableMetadataUnlisted,
         FullValidation,
         Signed,
+        SymbolsPackage
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PackageType.cs]: nit: trailing comma

File: src/NuGet.Services.EndToEnd/Support/PackageProperties.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+        public PackageProperties(bool isSymbolsPackage, HashSet<string> indexedFiles)
+        {
+            IsSymbolsPackage = isSymbolsPackage;
+            IndexedFiles = indexedFiles;
+        }
+
+        public static PackageProperties Default()
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PackageProperties.cs]: Why have this if the default ctor is public?

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 11]
```diff
@@ -96,21 +109,21 @@ private async Task<Package> PreparePackagesAsync(PackageType requestedPackageTyp
...
 
                 // Push all of the package types that have not been pushed yet.
-                var pushTasks = new Dictionary<PackageType, Task<Package>>();
+                var pushTasks = new Dictionary<PackageType, Task<List<Package>>>();
                 lock (_packagesLock)
                 {
                     foreach (var packageType in packageTypes)
                     {
-                        if (!_packages.TryGetValue(packageType, out Package package))
+                        if (!_packages.TryGetValue(packageType, out List<Package> package))
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: `IReadOnlyList<Package>`

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 12]
```diff
@@ -234,40 +282,132 @@ private PackageToPrepare InitializePackage(PackageType packageType)
...
+                    throw new Exception($"Error building symbols package! Error: {buildCommandResult.Error} Output: {buildCommandResult.Output}");
+                }
+
+                // Build nupkg and snupkg from appropriate files
+                var buildOutput = Path.Combine(testDirectory.FullPath, "bin", "Debug");
+                var dllFiles = Directory.GetFiles(buildOutput, "*.dll");
+                var pdbFiles = Directory.GetFiles(buildOutput, "*.pdb");
+                if (dllFiles.Count() == 0 || pdbFiles.Count() == 0)
+                {
+                    throw new Exception($"Symbols project build failed to create DLL or PDBs");
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Don't throw `Exception`. `InvalidOperationException` is better in this case.

An `Exception` cannot be caught. in a useful way.

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 13]
```diff
@@ -234,40 +282,132 @@ private PackageToPrepare InitializePackage(PackageType packageType)
...
+
+        private async Task<List<PackageToPrepare>> PrepareSymbolsPackageAsync(string id, string version, ITestOutputHelper logger)
+        {
+            using (var testDirectory = TestDirectory.Create())
+            {
+                // Copy the symbols project from template, also set the ID, Version appropriately in the properties.
+                var projectPath = CopySymbolsProjectFromTemplate(id, version, testDirectory.FullPath);
+
+                // Build the symbols project with the DotNet
+                var buildCommandResult = await DotNetExeClient.BuildProject(projectPath, logger);
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: `BuildProjectAsync`

File: src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj [hunk 14]
```diff
@@ -9,8 +9,9 @@
     <AppDesignerFolder>Properties</AppDesignerFolder>
     <RootNamespace>NuGet.Services.EndToEnd</RootNamespace>
     <AssemblyName>NuGet.Services.EndToEnd</AssemblyName>
-    <TargetFrameworkVersion>v4.5.2</TargetFrameworkVersion>
+    <TargetFrameworkVersion>v4.6</TargetFrameworkVersion>
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj]: We moved to 4.6.2 most places. Can we do that?

File: src/NuGet.Services.EndToEnd/Support/Clients/SymbolServerClient.cs (hunk 7 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/Clients/SymbolServerClient.cs]: I'll try to find documentation, not sure if there is. But this is the format for fetching portable pdbs from symbol server. I'll check with @cristinamanum.

File: src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+                throw new ArgumentNullException(nameof(pdbFullPath));
+            }
+
+            using (var stream = File.OpenRead(pdbFullPath))
+            using (var pdbReaderProvider = MetadataReaderProvider.FromPortablePdbStream(stream, MetadataStreamOptions.LeaveOpen))
+            {
+                var pdbReader = pdbReaderProvider.GetMetadataReader();
+                var pdbSignature = new BlobContentId(pdbReader.DebugMetadataHeader.Id).Guid;
+                var pdbFileName = Path.GetFileName(pdbFullPath).ToLowerInvariant();
+                var pdbAge = "FFFFFFFF";
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs]: >FFFFFFFF [](start = 30, length = 8)

This should be "ffffffff" I believe

File: src/NuGet.Services.EndToEnd/Support/TestData/E2E.TestPortableSymbols/Properties/AssemblyInfo.cs [hunk 16]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+// General Information about an assembly is controlled through the following
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("{0}")]
+[assembly: AssemblyDescription("")]
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/TestData/E2E.TestPortableSymbols/Properties/AssemblyInfo.cs]: Nit remove as unnecessary stuff as possible.

File: src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs (hunk 4 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/Clients/GalleryClient.cs]: I agree, but I would rather not make unrelated changes in this PR.

File: src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs (hunk 15 above)
Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs]: Here's the latest implementation on BaGet: [https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Core/Services/SymbolIndexingService.cs#L151](https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Core/Services/SymbolIndexingService.cs#L151)

I forget offhand what was broken with the uppercasing

---
In reply to: [241852502](https://github.com/NuGet/NuGet.Services.EndToEnd/pull/65#discussion_r241852502) [](ancestors = 241852502)

File: src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+using System.IO;
+using System.Reflection.Metadata;
+
+namespace NuGet.Services.EndToEnd.Support.Utilities
+{
+    public static class PortableMetadataReader
+    {
+        public static string GetIndex(string pdbFullPath)
+        {
+            var indexingPath = string.Empty;
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs]: nit set this after argument validation

File: src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+            }
+
+            using (var stream = File.OpenRead(pdbFullPath))
+            using (var pdbReaderProvider = MetadataReaderProvider.FromPortablePdbStream(stream, MetadataStreamOptions.LeaveOpen))
+            {
+                var pdbReader = pdbReaderProvider.GetMetadataReader();
+                var pdbSignature = new BlobContentId(pdbReader.DebugMetadataHeader.Id).Guid;
+                var pdbFileName = Path.GetFileName(pdbFullPath).ToLowerInvariant();
+                var pdbAge = "FFFFFFFF";
+                indexingPath = $"{pdbFileName}/{pdbSignature.ToString("N")}{pdbAge}/{pdbFileName}";
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs]: >ToString [](start = 61, length = 8)

I believe this should be `ToUpperInvariant()`. See: [https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Core/Services/SymbolIndexingService.cs#L161](https://github.com/loic-sharma/BaGet/blob/master/src/BaGet.Core/Services/SymbolIndexingService.cs#L161)

File: src/NuGet.Services.EndToEnd/Support/PackageType.cs (hunk 9 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/PackageType.cs]: not much of a fan of ending in comma, but don't feel too strongly.. added.

File: src/NuGet.Services.EndToEnd/Support/PackageProperties.cs (hunk 10 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/PackageProperties.cs]: removed.

File: src/NuGet.Services.EndToEnd/Support/PackageType.cs (hunk 9 above)
Comment (joelverhagen) [reply src/NuGet.Services.EndToEnd/Support/PackageType.cs]: it is why this diff is very clear 😄

File: src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs (hunk 15 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/Utilities/PortableMetadataReader.cs]: I don't think the casing matters here. I tried out all the casings and it worked fine when hitting the symbol server endpoint.

File: src/NuGet.Services.EndToEnd/Support/Package.cs [hunk 19]
```diff
@@ -61,6 +72,30 @@ public static Package Create(PackageCreationContext context)
...
+                .Select(x => new PhysicalPackageFile(x.Stream) {
+                    TargetPath = $"lib/{x.FileName}"
+                });
+
+            ReadOnlyCollection<byte> nupkgBytes;
+            using (var nupkgStream = TestData.BuildPackageStreamForFiles(context, physicalPackageFilesList, properties.IsSymbolsPackage))
+            {
+                var bufferStream = new MemoryStream();
+                nupkgStream.CopyTo(bufferStream);
+                nupkgBytes = Array.AsReadOnly(bufferStream.ToArray());
```
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Package.cs]: You don't need to do this. `ToArray()` copies the internal buffer.

File: src/NuGet.Services.EndToEnd/Support/Package.cs (hunk 1 above)
Comment (joelverhagen) [inline src/NuGet.Services.EndToEnd/Support/Package.cs]: `CopyTo` here too

Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/Package.cs]: done.

File: src/NuGet.Services.EndToEnd/Support/Package.cs (hunk 19 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/Package.cs]: caller needs readonly protection.

File: src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj (hunk 14 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/NuGet.Services.EndToEnd.csproj]: done.

File: src/NuGet.Services.EndToEnd/SymbolsPackageTests.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+            await _clients.SymbolServerClient.VerifySymbolFilesAvailable(
+                symbolsPackage.Id,
+                symbolsPackage.NormalizedVersion,
+                symbolsPackage.Properties.IndexedFiles,
+                _logger);
+        }
+
+        public void AssertPackageIsSymbolsPackage(Package package)
+        {
+            // The package should have indexed files if its a symbols package
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/SymbolsPackageTests.cs]: >its [](start = 56, length = 3)

it's

File: src/NuGet.Services.EndToEnd/Support/Package.cs [hunk 21]
```diff
@@ -61,6 +72,30 @@ public static Package Create(PackageCreationContext context)
...
+            var physicalPackageFilesList = files
+                .Select(x => new { FileName = Path.GetFileName(x), Stream = GetMemoryStreamForFile(x) })
+                .Select(x => new PhysicalPackageFile(x.Stream) {
+                    TargetPath = $"lib/{x.FileName}"
+                });
+
+            ReadOnlyCollection<byte> nupkgBytes;
+            using (var nupkgStream = TestData.BuildPackageStreamForFiles(context, physicalPackageFilesList, properties.IsSymbolsPackage))
+            {
+                var bufferStream = new MemoryStream();
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/Package.cs]: >bufferStream [](start = 20, length = 12)

Considering that the `bufferStream` is copied into `nupkgBytes`, shouldn't you dispose this stream?

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 22]
```diff
@@ -96,21 +110,21 @@ private async Task<Package> PreparePackagesAsync(PackageType requestedPackageTyp
...
                     }
                 }
 
                 await Task.WhenAll(pushTasks.Values);
-                
-                // Use the package that we just pushed.
+
+                // Use the packages that we just pushed.
                 return _packages[requestedPackageType];
             }
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: Shouldn't this line be protected by the `_packagesLock`? Someone else could be mutating `_packages` here

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs [hunk 23]
```diff
@@ -210,14 +253,20 @@ private IEnumerable<PackageType> GetPackageTypes(PackageType requestedPackageTyp
                 .OrderBy(x => x);
         }
 
-        private PackageToPrepare InitializePackage(PackageType packageType)
+        /// <summary>
+        /// Return the list of <see cref="PackageToPrepare"/> in the ordered form
+        /// for which we want to run all the tasks synchronously
```
Comment (loic-sharma) [inline src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: >synchronously [](start = 51, length = 13)

`synchronously` -> `sequentially`

File: src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs (hunk 22 above)
Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: This is an existing code path, and I don't think it needs lock, the lock and the logic before ensures that the _packages is writable only once for the corresponding `requestedPackageType`.

Comment (loic-sharma) [reply src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: You cannot read a dictionary as it is being mutated, regardless of the key

Comment (shishirx34) [reply src/NuGet.Services.EndToEnd/Support/PushedPackagesFixture.cs]: The logic ensures that the dictionary mutation happens before any thread could possibly read it here. However, it doesn't seem to be protected by any mutation from outside of this method(we don't have any today). I'll add the lock for possible future degradation.

**Pull request merged**